### ISSUE TITLE
Rule proposal - Naming rule - Member name should not contain parent type name

### ISSUE THREAD

**Issue by Przemyslaw-W** [DotNetAnalyzers/StyleCopAnalyzers#757]
Perhaps this rule should be: Method names should not include the class name.

Originally proposed by Andy Reeves: http://stylecop.codeplex.com/workitem/6883

---------------------------------------------------------

follow up comments:

tdyen wrote Jan 5, 2013 at 12:14 PM
By this do you mean 
```
public class List
{
    void ClearList()
    {
        // implementation here.
    }
}
```
But what about examples like
```
public class Type
{
    public static Type GetType(object parameter);
}
```
and
```
public class Table
{
    public static DataTable ToDataTable();
}
````

While your rule sounds mostly valid and quite sensible, I'm not sure its enforceable in all cases.

**Comment by Przemyslaw-W** [comment]
I like this rule. IMHO it has a lot of merit and targets design issue junior devs hit very often.

**But naming is hard.**
What I am concerned about, rule comes with many caveats. From top of my head:
* it should do comparison taking camel casing into account, so that method ```Broom()``` is not reported for class ```Room```
* it should strip leading "I" when checking interface members
* it should detect plural form is embedded in member name (not always simple -s suffix) so probably some kind of directory support might be needed
* if one writes code in other languages, this rule have limited use. E.g. in Polish nouns have singular and plural form, every comes with 7 grammar cases. Some of them overlap, but still there are potentially 14 forms to check for. Don't know how and if dictionaries support such stuff.

Even with above tweaks, it probably will produce ton of false positives, which should be suppressed, causing a lot of work for dev. If we agree to ad such rule, I think we should run this across few big open source projects, including roslyn and net core. This should allow us to detect and handle more "special" cases and see if such rule could be released into wild.

Also it might turn out, such rule might be better suited for other library, not necessary StyleCop

**Comment by sharwell** [comment]
I'm not sure this rule would have the intended effect. For example, `Enumerable.AsEnumerable` would give a warning, even though it's a recognized and easily understandable method. At least for now I vote no.

**Comment by vweijsters** [comment]
This is a nice example of something that's easy to spot for humans, but extremely difficult to get right in software. Besides that I think it might be perceived as patronizing by lots of people, so I will vote no for this one.

**Comment by GregReddick** [comment]
I think I have many places of legitimate code that would trigger this. :-1:
